book/course/spack/advanced.md

```diff
@@ -0,0 +1,194 @@
+individual package, but also you can tweak the dependency decisions, to use
+different packages if you know better than the defaults.
+
+## Variants
+
+Variants are where the package has options how to build it.  Let's pick on
+`piranha` as a nice example.  You can get information on a package in Spack,
+which will tell you about how it can be built:
+
+```bash
+$ spack info piranha
+CMakePackage:   piranha
+
+Description:
+    Piranha is a computer-algebra library for the symbolic manipulation of
+    sparse multivariate polynomials and other closely-related symbolic
+    objects (such as Poisson series).
+
+Homepage: https://bluescarni.github.io/piranha/sphinx/
+
+Preferred version:
+    0.5        https://github.com/bluescarni/piranha/archive/v0.5.tar.gz
+
+Safe versions:
+    develop    [git] https://github.com/bluescarni/piranha.git on branch master
+    0.5        https://github.com/bluescarni/piranha/archive/v0.5.tar.gz
+
+Deprecated versions:
+    None
+
+Variants:
+    Name [Default]                 When    Allowed values          Description
+    ===========================    ====    ====================    ==================================
+
+    build_type [RelWithDebInfo]    --      Debug, Release,         CMake build type
+                                           RelWithDebInfo,
+                                           MinSizeRel
+    ipo [off]                      --      on, off                 CMake interprocedural optimization
+    python [on]                    --      on, off                 Build the Python bindings
+
+Build Dependencies:
+    boost  bzip2  cmake  gmp  mpfr  python
+
+Link Dependencies:
+    boost  bzip2  gmp  mpfr
+
+Run Dependencies:
+    python
+```
+
+From this you can see that there's lots of versions available, and there's a
+number of variants listed, along with dependencies.  Let's say I wanted to
+build the develop version, and we want ipo to be on, python off, and a
+build\_type of Release.  Basically, be as awkward as possible and change
+everything:
+
+```bash
+spack install piranha @develop +ipo -python build_type=Release
+```
+
+There we're just turned on a feature (ipo), turned off a feature (python), and
+told it we want the Release build type.  Have a look through the references to
+get further information, but this is actually the basics of choosing variants
+of packages.
+
+## Compilers
+
+You may want to build a package with a different compiler.  Let's say I have
+gcc@11.2.1 available:
+
+```bash
+spack install piranha @develop +ipo -python build_type=Release %gcc@11.2.1
+```
+
+Now Spack will happily rebuild that specific variant you've asked for, with the
+compiler you've asked for.
+
+## Dependencies
+
+You can also specify particular versions and variants of dependencies.  We can
+take our previous build even further:
+
+```bash
+spack install piranha @develop +ipo -python build_type=Release ^boost@1.62.0+thread %gcc@11.2.1
```

Comment:
Can you clarify the distinction of using the `^` for boost but the `%` for gcc in the text?
My assumption is package v compiler but might be worth clarifying.